src/hb-ot-color-colr-table.hh

```diff
@@ -39,6 +40,64 @@
 
 namespace OT {
 
+struct hb_colrv1_closure_context_t :
+       hb_dispatch_context_t<hb_colrv1_closure_context_t>
+{
+  template <typename T>
+  return_t dispatch (const T &obj) { obj.closurev1 (this); return hb_empty_t (); }
+  static return_t default_return_value () { return hb_empty_t (); }
+
+  bool paint_visited (const void *paint)
+  {
+    hb_codepoint_t delta = (hb_codepoint_t) ((uintptr_t) paint - (uintptr_t) base);
+     if (visited_paint->has (delta))
+      return true;
+
+    visited_paint->add (delta);
+    return false;
+  }
+
+  bool should_visit_paint (const void *paint)
+  {
+    if (cycle_detected) return false;
+    if (paint_visited (paint))
+    {
+      cycle_detected = true;
```

Comment:
Since this is a directed graph re-visiting the same node doesn't imply that there's a cycle, for that you need to look for back edges (https://en.wikipedia.org/wiki/Cycle_(graph_theory)#Cycle_detection). That said I don't think we need to explicitly detect cycles, rather just avoid re-visiting the same nodes (so it doesn't infinite loop when there are cycles). So you can drop the cycle_detected boolean and just stick with the visited set.